lua/bson_lua_wrapper.c

```diff
@@ -0,0 +1,372 @@
+        entry = bson_object_iterator_next(&iter);
+    }
+    return 0;
+}
+
+static int bson_array_to_table(lua_State *L, BsonArray *arr, char *errorMessage) {
+    lua_newtable(L); //Stack: [table]
+    int index;
+    for (index = 0; index <= arr->count; index++) {
+        BsonElement *element = bson_array_get(arr, index);
+        
+        int success = bson_element_to_table(L, element, errorMessage); //Stack [table, {type, value}]
+        if (success != 0) {
+            return success;
+        }
+
+        lua_rawseti(L, -2, index + 1 /*Lua indexing*/); //Stack [table]
+    }
+    return 0;
+}
+
+static int bson_element_to_table(lua_State *L, BsonElement *element, char *errorMessage) {
+    lua_newtable(L); //Stack: [{}];
+
+    lua_pushinteger(L, ((lua_Integer)element->type) & 0xFF); //Stack: [{}, type]
+    lua_setfield(L, -2, "type"); //Stack: [{type}]
+
+    switch (element->type) {
+        case TYPE_DOCUMENT: {
+            BsonObject *value = (BsonObject *)element->value;
+            int success = bson_object_to_table(L, value, errorMessage); //Stack [{type}, subTable]
+            if (success != 0) {
+                return success;
+            }
+            break;
+        }
+        case TYPE_ARRAY: {
+            BsonArray *value = (BsonArray *)element->value;
+            int success = bson_array_to_table(L, value, errorMessage); //Stack [{type}, subTable]
+            if (success != 0) {
+                return success;
+            }
+            break;
+        }
+        case TYPE_INT32: {
+            int32_t value = *(int32_t *)element->value;
+            lua_pushinteger(L, (lua_Integer)value); //Stack [{type}, value]
+            break;
+        }
+        case TYPE_INT64: {
+            int64_t value = *(int64_t *)element->value;
+            lua_pushinteger(L, (lua_Integer)value); //Stack [{type}, value]
+            break;
+        }
+        case TYPE_STRING: {
+            char *value = (char *)element->value;
+            lua_pushstring(L, value); //Stack [{type}, value]
+            break;
+        }
+        case TYPE_DOUBLE: {
+            double value = *(double *)element->value;
+            lua_pushnumber(L, (lua_Number)value); //Stack [{type}, value]
+            break;
+        }
+        case TYPE_BOOLEAN: {
+            bson_boolean value = *(bson_boolean *)element->value; //Stack [{type}, value]
+            lua_pushboolean(L, (int)value);
+            break;
+        }
+        default: {
+            snprintf(errorMessage, 255, "Unsupported  type in BSON object: %i", (int)element->type);
+            return -1;
+        }
+    }
+    lua_setfield(L, -2, "value"); //Stack [{type, value}]
+    return 0;
+}
+
+static const struct luaL_Reg bson_object_methods[] = {
+    { "to_bytes", lbson_object_to_bytes },
+    { "to_table", lbson_object_from_bytes },
+    { NULL, NULL }
+};
+
+int luaopen_bson4lua(lua_State *L) {
+    // Create metatable
+    luaL_newmetatable(L, "BSONObject"); //Stack: [{}]
```

Comment:
@jacobkeeler Could you please describe purpose of creation of current metatable?

Reply:
Remnant of something else I was trying to do, I will remove